File: SIN BDD/SOAP_JAVA_CONUNI_GR06/05. CLIESCRITORIO/cliente_escritorio_soap_java_conuni_gr06/src/ec/edu/monster/controlador/ControladorEscritorio.java
package ec.edu.monster.controlador;

import ec.edu.monster.modelo.FormatoConversion;
import ec.edu.monster.modelo.Resultado;
import ec.edu.monster.modelo.ServicioAutenticacion;
import ec.edu.monster.modelo.ServicioLongitud;
import ec.edu.monster.modelo.ServicioMasa;
import ec.edu.monster.modelo.ServicioTemperatura;
import ec.edu.monster.vista.PanelConversion;
import ec.edu.monster.vista.VentanaPrincipal;
import javax.swing.SwingWorker;

/**
 * Controlador de la aplicacion de escritorio. Cablea los paneles de la Vista
 * con los servicios del Modelo. Usa {@link SwingWorker} para que las llamadas
 * SOAP no bloqueen el hilo de eventos (Event Dispatch Thread) y la UI se
 * mantenga responsiva.
 */
public class ControladorEscritorio {

    private final VentanaPrincipal ventana;

    private final ServicioAutenticacion servicioAutenticacion = new ServicioAutenticacion();
    private final ServicioLongitud servicioLongitud = new ServicioLongitud();
    private final ServicioMasa servicioMasa = new ServicioMasa();
    private final ServicioTemperatura servicioTemperatura = new ServicioTemperatura();

    private String usuarioActual;

    public ControladorEscritorio() {
        this.ventana = new VentanaPrincipal();
        cablearVista();
    }

    public void iniciar() {
        ventana.mostrar(VentanaPrincipal.TARJETA_LOGIN);
        ventana.setVisible(true);
    }

    // ========= Cableado =========

    private void cablearVista() {
        ventana.getPanelLogin().setOnLogin(this::manejarLogin);
        ventana.getPanelMenu().setOnCategoriaSeleccionada(this::manejarCategoria);
        ventana.getPanelMenu().setOnCerrarSesion(this::manejarCerrarSesion);
        ventana.getPanelConversion().setOnConvertir(this::manejarConvertir);
        ventana.getPanelConversion().setOnVolver(this::manejarVolverAlMenu);
    }

    // ========= Autenticacion =========

    private void manejarLogin(String usuario, String contrasena) {
        if (usuario.isEmpty() || contrasena.isEmpty()) {
            ventana.getPanelLogin().mostrarError("Completa usuario y contraseña");
            return;
        }
        ventana.getPanelLogin().mostrarError(" ");
        ventana.getPanelLogin().setBotonHabilitado(false);

        new SwingWorker<Boolean, Void>() {
            private String mensajeError;

            @Override
            protected Boolean doInBackground() {
                try {
                    return servicioAutenticacion.iniciarSesion(usuario, contrasena);
                } catch (Exception ex) {
                    mensajeError = "No se pudo conectar: " + ex.getMessage();
                    return false;
                }
            }

            @Override
            protected void done() {
                ventana.getPanelLogin().setBotonHabilitado(true);
                try {
                    boolean ok = get();
                    if (mensajeError != null) {
                        ventana.getPanelLogin().mostrarError(mensajeError);
                        return;
                    }
                    if (ok) {
                        usuarioActual = usuario;
                        ventana.getPanelMenu().setUsuario(usuario);
                        ventana.getPanelLogin().limpiar();
                        ventana.mostrar(VentanaPrincipal.TARJETA_MENU);
                    } else {
                        ventana.getPanelLogin().mostrarError("Usuario o contraseña inválidos");
                    }
                } catch (Exception ex) {
                    ventana.getPanelLogin().mostrarError("Error inesperado: " + ex.getMessage());
                }
            }
        }.execute();
    }

    private void manejarCerrarSesion() {
        usuarioActual = null;
        ventana.mostrar(VentanaPrincipal.TARJETA_LOGIN);
    }

    // ========= Navegacion al conversor =========

    private void manejarCategoria(String categoria) {
        ventana.getPanelConversion().setCategoria(categoria);
        ventana.mostrar(VentanaPrincipal.TARJETA_CONVERSOR);
    }

    private void manejarVolverAlMenu() {
        ventana.mostrar(VentanaPrincipal.TARJETA_MENU);
    }

    // ========= Conversiones =========

    private void manejarConvertir(String operacion, Double valor) {
        PanelConversion panel = ventana.getPanelConversion();
        String categoria = panel.getCategoria();
        panel.setBotonHabilitado(false);

        new SwingWorker<Resultado, Void>() {
            @Override
            protected Resultado doInBackground() {
                return ejecutar(categoria, operacion, valor);
            }

            @Override
            protected void done() {
                panel.setBotonHabilitado(true);
                try {
                    panel.mostrarResultado(get());
                } catch (Exception ex) {
                    panel.mostrarResultado(Resultado.error("Error: " + ex.getMessage()));
                }
            }
        }.execute();
    }

    private Resultado ejecutar(String categoria, String operacion, double valor) {
        try {
            double r;
            switch (categoria) {
                case "longitud":
                    r = llamarLongitud(operacion, valor);
                    break;
                case "masa":
                    r = llamarMasa(operacion, valor);
                    break;
                case "temperatura":
                    r = llamarTemperatura(operacion, valor);
                    break;
                default:
                    return Resultado.error("Categoría desconocida");
            }
            String[] u = FormatoConversion.unidades(operacion);
            return Resultado.ok(FormatoConversion.formatear(valor, u[0], r, u[1]));
        } catch (Exception ex) {
            return Resultado.error("Error del servicio: " + ex.getMessage());
        }
    }

    private double llamarLongitud(String op, double v) throws Exception {
        switch (op) {
            case "metrosAPies":          return servicioLongitud.metrosAPies(v);
            case "kilometrosAMillas":    return servicioLongitud.kilometrosAMillas(v);
            case "centimetrosAPulgadas": return servicioLongitud.centimetrosAPulgadas(v);
            case "yardasAMetros":        return servicioLongitud.yardasAMetros(v);
            case "milimetrosAPulgadas":  return servicioLongitud.milimetrosAPulgadas(v);
            default: throw new IllegalArgumentException("Operación inválida: " + op);
        }
    }

    private double llamarMasa(String op, double v) throws Exception {
        switch (op) {
            case "kilogramosALibras":    return servicioMasa.kilogramosALibras(v);
            case "gramosAOnzas":         return servicioMasa.gramosAOnzas(v);
            case "toneladasAKilogramos": return servicioMasa.toneladasAKilogramos(v);
            case "librasAOnzas":         return servicioMasa.librasAOnzas(v);
            case "miligramosAGramos":    return servicioMasa.miligramosAGramos(v);
            default: throw new IllegalArgumentException("Operación inválida: " + op);
        }
    }

    private double llamarTemperatura(String op, double v) throws Exception {
        switch (op) {
            case "celsiusAFahrenheit": return servicioTemperatura.celsiusAFahrenheit(v);
            case "fahrenheitACelsius": return servicioTemperatura.fahrenheitACelsius(v);
            case "celsiusAKelvin":     return servicioTemperatura.celsiusAKelvin(v);
            case "kelvinACelsius":     return servicioTemperatura.kelvinACelsius(v);
            case "fahrenheitAKelvin":  return servicioTemperatura.fahrenheitAKelvin(v);
            default: throw new IllegalArgumentException("Operación inválida: " + op);
        }
    }
}

File: SIN BDD/SOAP_DOTNET_CONUNI_GR06/05. CLIESCRITORIO/cliente_escritorio_soap_dotnet_conuni_gr06/src/Controlador/ControladorEscritorio.cs
using Ec.Edu.Monster.Modelo;

namespace Ec.Edu.Monster.Controlador;

public class ControladorEscritorio
{
    private readonly ServicioAutenticacion autenticacion = new();
    private readonly ServicioLongitud longitud = new();
    private readonly ServicioMasa masa = new();
    private readonly ServicioTemperatura temperatura = new();

    public Resultado IniciarSesion(string usuario, string contrasena)
    {
        var exito = autenticacion.Autenticar(usuario, contrasena);
        return new Resultado
        {
            Exito = exito,
            Mensaje = exito ? "Sesion iniciada" : "Credenciales invalidas"
        };
    }

    public Resultado ConvertirLongitud(string operacion, double valor)
    {
        return operacion switch
        {
            "metrosAPies" => new Resultado { Exito = true, Valor = longitud.MetrosAPies(valor) },
            "kilometrosAMillas" => new Resultado { Exito = true, Valor = longitud.KilometrosAMillas(valor) },
            "centimetrosAPulgadas" => new Resultado { Exito = true, Valor = longitud.CentimetrosAPulgadas(valor) },
            "yardasAMetros" => new Resultado { Exito = true, Valor = longitud.YardasAMetros(valor) },
            "milimetrosAPulgadas" => new Resultado { Exito = true, Valor = longitud.MilimetrosAPulgadas(valor) },
            _ => new Resultado { Exito = false, Mensaje = "Operacion no valida" }
        };
    }

    public Resultado ConvertirMasa(string operacion, double valor)
    {
        return operacion switch
        {
            "kilogramosALibras" => new Resultado { Exito = true, Valor = masa.KilogramosALibras(valor) },
            "gramosAOnzas" => new Resultado { Exito = true, Valor = masa.GramosAOnzas(valor) },
            "toneladasAKilogramos" => new Resultado { Exito = true, Valor = masa.ToneladasAKilogramos(valor) },
            "librasAOnzas" => new Resultado { Exito = true, Valor = masa.LibrasAOnzas(valor) },
            "miligramosAGramos" => new Resultado { Exito = true, Valor = masa.MiligramosAGramos(valor) },
            _ => new Resultado { Exito = false, Mensaje = "Operacion no valida" }
        };
    }

    public Resultado ConvertirTemperatura(string operacion, double valor)
    {
        return operacion switch
        {
            "celsiusAFahrenheit" => new Resultado { Exito = true, Valor = temperatura.CelsiusAFahrenheit(valor) },
            "fahrenheitACelsius" => new Resultado { Exito = true, Valor = temperatura.FahrenheitACelsius(valor) },
            "celsiusAKelvin" => new Resultado { Exito = true, Valor = temperatura.CelsiusAKelvin(valor) },
            "kelvinACelsius" => new Resultado { Exito = true, Valor = temperatura.KelvinACelsius(valor) },
            "fahrenheitAKelvin" => new Resultado { Exito = true, Valor = temperatura.FahrenheitAKelvin(valor) },
            _ => new Resultado { Exito = false, Mensaje = "Operacion no valida" }
        };
    }
}
